File: Java-Core/src/main/java/presto/statement/RaiseStatement.java
package presto.statement;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.error.UserError;
import presto.expression.IExpression;
import presto.grammar.Identifier;
import presto.runtime.Context;
import presto.type.CategoryType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class RaiseStatement extends SimpleStatement {
	
	IExpression expression;
	
	public RaiseStatement(IExpression expression) {
		this.expression = expression;
	}

	public IExpression getExpression() {
		return expression;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
		case S:
			writer.append("raise ");
			break;
		case O:
			writer.append("throw ");
			break;
		}
		expression.toDialect(writer);
	}
	
	@Override
	public String toString() {
		return "raise " + expression.toString();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj==this)
			return true;
		if(obj==null)
			return false;
		if(!(obj instanceof RaiseStatement))
			return false;
		RaiseStatement other = (RaiseStatement)obj;
		return this.getExpression().equals(other.getExpression());
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType type = expression.check(context);
		if(!type.isAssignableTo(context, new CategoryType(new Identifier("Error"))))
			throw new SyntaxError(type.getName() + " does not extend Error");
		return type;
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		throw new UserError(expression);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/statement/RaiseStatement.js
var SimpleStatement = require("./SimpleStatement").SimpleStatement;
var CategoryType = require("../type/CategoryType").CategoryType;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var UserError = require("../error/UserError").UserError;
var Dialect = require("../parser/Dialect").Dialect;

function RaiseStatement(expression) {
	SimpleStatement.call(this);
	this.expression = expression;
	return this;
}

RaiseStatement.prototype = Object.create(SimpleStatement.prototype);
RaiseStatement.prototype.constructor = RaiseStatement;

RaiseStatement.prototype.toString = function() {
	return "raise " + this.expression.toString();
};

RaiseStatement.prototype.equals = function(obj) {
	if(obj==this) {
		return true;
	} else if(!(obj instanceof RaiseStatement)) {
		return false;
	} else {
		return this.expression.equals(other.expression);
	}
};

RaiseStatement.prototype.check = function(context) {
	var type = this.expression.check(context);
	if(!type.isAssignableTo(context, new CategoryType("Error"))) {
		throw new SyntaxError(type.name + " does not extend Error");
	}
	return type;
};

RaiseStatement.prototype.interpret = function(context) {
	throw new UserError(this.expression);
};

RaiseStatement.prototype.toDialect = function(writer) {
    switch(writer.dialect) {
        case Dialect.E:
        case Dialect.S:
            writer.append("raise ");
            break;
        case Dialect.O:
            writer.append("throw ");
            break;
    }
    this.expression.toDialect(writer);
};

exports.RaiseStatement = RaiseStatement;
